apps/server/lib/http/auth.ts

```diff
@@ -0,0 +1,81 @@
+import {
+	Context,
+	JellyfishKernel,
+	SessionContract,
+} from '@balena/jellyfish-types/build/core';
+import bcrypt from 'bcrypt';
+import _ from 'lodash';
+
+const isValidToken = async (
+	context: Context,
+	jellyfish: JellyfishKernel,
+	session: string,
+	sessionToken: string = '',
+): Promise<boolean> => {
+	try {
+		const card = await jellyfish.getCardById<SessionContract>(
+			context,
+			jellyfish.sessions!.admin,
+			session,
+		);
+		return bcrypt.compare(sessionToken, card!.data.token!.authentication);
+	} catch (e) {
+		return false;
+	}
+};
+
+const isValidSession = async (
+	context: Context,
+	jellyfish: JellyfishKernel,
+	session: string,
+): Promise<boolean> => {
+	try {
+		const card = await jellyfish.getCardById<SessionContract>(
+			context,
+			jellyfish.sessions!.admin,
+			session,
+		);
+		if (!card) {
+			return false;
+		}
+		return true;
+	} catch (e) {
+		return false;
+	}
+};
+
+export const authMiddleware =
+	(jellyfish: JellyfishKernel, options: { guestSession: string }) =>
+	async (request, _response, next) => {
+		const authorizationHeader = request.headers['authorization'];
+
+		// Split 'Bearer ' and <session_id>.<auth_token>
+		const authorizationHeaderValue = _.last(_.split(authorizationHeader, ' '));
+
+		// Split session ID and auth token
+		const credentials = _.split(authorizationHeaderValue, '.');
+
+		request.session = options.guestSession;
+
+		if (credentials.length === 2) {
+			const sessionIsValid = await isValidToken(
+				request.context,
+				jellyfish,
+				credentials[0],
+				credentials[1],
+			);
+			if (sessionIsValid) {
+				request.session = credentials[0];
+			}
+		} else if (credentials.length === 1) {
```

Comment:
@edufschmidt Did some local testing and found that this path is taken even when an `Authorization` header isn't set, for example on the initial login when executing `action-create-session`. This results in an unnecessary call to the database that in turn results in a Postgres error:
```
   47 export const authMiddleware =
   48   (jellyfish: JellyfishKernel, options: { guestSession: string }) =>
>> 49   async (request, _response, next) => {
+  50     console.log('=== authMiddleware');
   51     const authorizationHeader = request.headers['authorization'];
   52 
   53     // Split 'Bearer ' and <session_id>.<auth_token>
   54     const authorizationHeaderValue = _.last(_.split(authorizationHeader, ' '));
   55 
   56     // Split session ID and auth token
   57     const credentials = _.split(authorizationHeaderValue, '.');
+  58     console.log('credentials:', credentials);
   59 
   60     request.session = options.guestSession;
   61 
   62     if (credentials.length === 2) {
   63       const sessionIsValid = await isValidToken(
   64         request.context,
   65         jellyfish,
   66         credentials[0],
   67         credentials[1],
   68       );
   69       if (sessionIsValid) {
   70         request.session = credentials[0];
   71       }
   72     } else if (credentials.length === 1) {
+  73       console.log('credentials[0]:', credentials[0]);
   74       const sessionIsValid = await isValidSession(
   75         request.context,
   76         jellyfish,
   77         credentials[0],
   78       );
   79       if (sessionIsValid) {
   80         request.session = credentials[0];
   81       }
   82     }
   83     return next();
   84   };
```

Logs:
```
[Logs]    [1/11/2022, 11:56:39 AM] [api] === authMiddleware
[Logs]    [1/11/2022, 11:56:39 AM] [api] credentials: [ '' ]
[Logs]    [1/11/2022, 11:56:39 AM] [api] credentials[0]: 
[Logs]    [1/11/2022, 11:56:39 AM] [postgres] 2022-01-11 02:56:39.129 UTC [309] ERROR:  invalid input syntax for type uuid: "" at character 280
[Logs]    [1/11/2022, 11:56:39 AM] [postgres] 2022-01-11 02:56:39.129 UTC [309] STATEMENT:  SELECT to_jsonb(cards) || jsonb_build_object('version', CONCAT_WS('+', CONCAT_WS('-', CONCAT_WS('.', cards.version_major, cards.version_minor, cards.version_patch), NULLIF(cards.version_prerelease, '') ), NULLIF(cards.version_build, '') )) AS payload
[Logs]    [1/11/2022, 11:56:39 AM] [postgres] 	FROM cards
[Logs]    [1/11/2022, 11:56:39 AM] [postgres] 	WHERE cards.id = ''
[Logs]    [1/11/2022, 11:56:39 AM] [postgres] 	LIMIT 1
```

Reply:
Thanks for pointing it out @joshbwlng. I patched the auth middleware to account for the case in which the authorization header is missing in the request. I also tested my changes locally, and the Postgres error is gone.